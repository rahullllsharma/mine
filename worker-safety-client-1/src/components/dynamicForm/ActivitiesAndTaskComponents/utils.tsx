import { ActionLabel, BodyText } from "@urbint/silica";
import { Checkbox } from "../../forms/Basic/Checkbox/Checkbox";
import { RiskLevel } from "../../riskBadge/RiskLevel";
import { CWFTaskCard } from "./CWFTaskCard";

const StaticTaskCard = () => {
  return (
    <div className="mt-2.5">
      <Checkbox
        className="w-full gap-4"
        checked={true}
        disabled={true}
        onClick={() => void 0}
      >
        <ActionLabel className="mt-2.5">Activity Title</ActionLabel>
      </Checkbox>
      <Checkbox
        className="w-full gap-4 mt-2.5"
        checked={true}
        disabled={true}
        labelClassName="w-full"
        onClick={() => void 0}
      >
        <CWFTaskCard
          title={"Task Title"}
          risk={RiskLevel.UNKNOWN}
          showRiskInformation={true}
          showRiskText={true}
          withLeftBorder={true}
        />
      </Checkbox>
    </div>
  );
};

const TasksBlankField = ({ errorMessage }: { errorMessage?: string }) => {
  return (
    <div className="flex flex-col items-center sm:flex-row gap-2 sm:gap-4 p-2 sm:p-4 bg-gray-100 h-24 sm:h-32 w-full mt-5">
      <BodyText className="text-neutrals-secondary">
        There are currently no Activities or Tasks for the selected Work Dates
        and Times.
        {errorMessage && (
          <>
            <div className="text-system-error-40 text-sm pt-2 block text-center">
              {errorMessage}
            </div>
          </>
        )}
      </BodyText>
    </div>
  );
};

export { StaticTaskCard, TasksBlankField };
