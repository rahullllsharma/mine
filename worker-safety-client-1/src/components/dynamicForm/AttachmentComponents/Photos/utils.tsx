import type {
  DownloadResult,
  PhotoUploadItem,
  SubmissionPhotoType,
  UploadItem,
} from "@/components/templatesComponents/customisedForm.types";
import { messages } from "@/locales/messages";
import {
  ALLOWED_DOCUMENTS_FORMATS,
  ALLOWED_IMAGE_FORMATS,
  maxFiles,
} from "@/utils/customisedFormUtils/customisedForm.constants";

const BlankField = () => {
  return (
    <div className="flex flex-col sm:flex-row gap-2 sm:gap-4 p-2 sm:p-4 bg-gray-100 h-24 sm:h-32 w-full" />
  );
};

const photosFormatsText = ALLOWED_IMAGE_FORMATS.join(",")
  .replace(/\,/g, ", ")
  .toUpperCase();

const documentFormatsText = ALLOWED_DOCUMENTS_FORMATS.join(",")
  .replace(/\,/g, ", ")
  .toUpperCase();

const getMaxFilesErrorMessage = (isPhoto: boolean): string => {
  const contentType = isPhoto ? "images" : "documents";
  return `Maximum of ${maxFiles} ${contentType} can be uploaded at once.`;
};
const convertToPhotoUploadItem = (
  submissionPhotos: SubmissionPhotoType[]
): PhotoUploadItem[] => {
  return submissionPhotos.map(photo => {
    const {
      id,
      url,
      date,
      name,
      size,
      time,
      last_modified: lastModified,
      signed_url: signedUrl,
      display_name: displayName,
      description = "",
    } = photo;
    return {
      id: id,
      url: url,
      date: date,
      name: name,
      size: size,
      time: time,
      signedUrl: signedUrl,
      lastModified: lastModified,
      displayName: displayName,
      description: description || "",
    };
  });
};

const convertToSubmissionPhotoType = (
  uploadedFiles: PhotoUploadItem[]
): SubmissionPhotoType[] => {
  return uploadedFiles.map(file => {
    const {
      id,
      url,
      date,
      name,
      size,
      time,
      signedUrl,
      lastModified,
      displayName,
      description = "",
    } = file;
    return {
      id: id,
      md5: null,
      url: url,
      date: date,
      name: name,
      size: size,
      time: time,
      crc32c: null,
      category: null,
      mimetype: null,
      signed_url: signedUrl,
      last_modified: lastModified,
      display_name: displayName,
      description: description || "",
    };
  });
};

const timeoutPromise = (ms: number) =>
  new Promise((_, reject) => {
    setTimeout(() => reject(new Error(messages.requestTimedOut)), ms);
  });

const download = async (
  path: string,
  name: string
): Promise<DownloadResult> => {
  let result: DownloadResult = "success";
  try {
    const file = await fetch(path);
    const fileBlob = await file.blob();
    const url = URL.createObjectURL(fileBlob);
    const link = document.createElement("a");
    link.href = url;
    link.download = name;
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
  } catch (error) {
    result = "error";
  }
  return result;
};

const downloadDocumentHandler = async (
  event: React.MouseEvent<HTMLButtonElement>,
  file: UploadItem
) => {
  event.preventDefault();
  await download(file.signedUrl, file.name);
};

const isImage = (fileName: string): boolean => {
  const imageExtensions = ALLOWED_IMAGE_FORMATS;
  const ext = fileName.split(".").pop()?.toLowerCase();
  return ext ? imageExtensions.includes(ext) : false;
};

export {
  BlankField,
  photosFormatsText,
  documentFormatsText,
  getMaxFilesErrorMessage,
  convertToPhotoUploadItem,
  convertToSubmissionPhotoType,
  timeoutPromise,
  download,
  downloadDocumentHandler,
  isImage,
};
