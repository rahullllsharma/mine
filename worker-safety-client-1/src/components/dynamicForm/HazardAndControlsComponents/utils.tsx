import { BodyText, CaptionText } from "@urbint/silica";
import BlankFieldTemplates from "../BlankFieldTemplates";
import Modal from "../../shared/modal/Modal";
import ButtonDanger from "../../shared/button/danger/ButtonDanger";
import ButtonRegular from "../../shared/button/regular/ButtonRegular";
import ButtonSecondary from "../../shared/button/secondary/ButtonSecondary";

interface HazardsCardBlankFieldProps {
  energyType?: string;
  subTitle?: string;
  addHazardHandler: (tab?: string) => void;
  readOnly?: boolean;
  inSummary?: boolean;
}
export type DeleteHazardModalProps = {
  isDeleteModalOpen: boolean;
  setIsDeleteModalOpen: (value: boolean) => void;
  setHazardToDelete: (value: string | null) => void;
  confirmDeleteHazard: () => void;
};

const HazardsBlankField = () => {
  return (
    <>
      <div className="text-xs text-neutral-600 mt-0.5 mb-4">
        <CaptionText>
          Select the tasks you were responsible for overseeing at this location.
        </CaptionText>
      </div>
      <BlankFieldTemplates />
    </>
  );
};

const HazardsCardBlankField = ({
  addHazardHandler,
  subTitle,
  energyType,
  readOnly,
  inSummary,
}: HazardsCardBlankFieldProps) => {
  const tab = energyType === "HIGH_ENERGY" ? "highEnergy" : "otherHazards";
  return (
    <div className="mb-4 mt-5 bg-gray-100 rounded p-3">
      <div className="flex items-center">
        <BodyText className="text-xl font-semibold mb-2">
          {energyType === "HIGH_ENERGY"
            ? subTitle || "High Energy Hazards"
            : "Other Hazards"}
        </BodyText>
        {!inSummary && (
          <ButtonSecondary
            label="Add Hazards"
            iconStart="plus_circle_outline"
            size="sm"
            className="mb-2 ml-auto"
            onClick={() => addHazardHandler(tab)}
            disabled={readOnly}
          />
        )}
      </div>
      <div className="bg-white border rounded-lg p-4 px-8">
        {energyType === "HIGH_ENERGY" ? (
          <>
            <CaptionText className="font-semibold text-center text-md text-neutrals-secondary">
              Identify and Control {subTitle || "High Energy Hazards"}
            </CaptionText>
            <CaptionText className="text-center text-md text-neutrals-secondary">
              {subTitle} can cause serious injury or fatality. Add all
              applicable hazards and select the appropriate controls to
              eliminate, reduce, or mitigate the exposure.
            </CaptionText>
          </>
        ) : (
          <>
            <CaptionText className="font-semibold text-center text-md text-neutrals-secondary">
              Identify and Control Other Hazards
            </CaptionText>
            <CaptionText className="text-center text-md text-neutrals-secondary">
              Other hazards can result in incidents or injuries that require
              medical attention. Add all applicable hazards and select the
              appropriate controls to eliminate, reduce, or mitigate the
              exposure.
            </CaptionText>
          </>
        )}
      </div>
    </div>
  );
};

const DeleteHazardModal = ({
  isDeleteModalOpen,
  setIsDeleteModalOpen,
  setHazardToDelete,
  confirmDeleteHazard,
}: DeleteHazardModalProps) => {
  return (
    <Modal
      title="Delete Hazard"
      isOpen={isDeleteModalOpen}
      closeModal={() => {
        setIsDeleteModalOpen(false);
        setHazardToDelete(null);
      }}
      size="md"
    >
      <div className="mb-10">
        <BodyText className="text-neutral-shade-75">
          Are you sure you want to delete this hazard?
        </BodyText>
        <BodyText className="text-neutral-shade-75">
          This action cannot be undone.
        </BodyText>
      </div>
      <div className="flex justify-end gap-3">
        <ButtonRegular
          label="Cancel"
          onClick={() => {
            setIsDeleteModalOpen(false);
            setHazardToDelete(null);
          }}
        />
        <ButtonDanger label={`Confirm`} onClick={confirmDeleteHazard} />
      </div>
    </Modal>
  );
};

export { HazardsBlankField, HazardsCardBlankField, DeleteHazardModal };
