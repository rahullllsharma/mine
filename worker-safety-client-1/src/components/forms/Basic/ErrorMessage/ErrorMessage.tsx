import type { FormField } from "@/utils/formField";
import type { Errors } from "io-ts";
import classnames from "classnames";
import { isLeft } from "fp-ts/Either";
import { failure } from "io-ts/PathReporter";

type ErrorMessageProps<V> = {
  className?: string;
  field: FormField<Errors, string, V>;
};

function ErrorMessage<V>({ className, field }: ErrorMessageProps<V>) {
  const { val } = field;
  return (
    <>
      {isLeft(val) &&
        failure(val.left).map(message => (
          <span
            key={message}
            className={classnames(
              "text-sm mt-1 text-system-error-40",
              className
            )}
          >
            {message}
          </span>
        ))}
    </>
  );
}

export { ErrorMessage };
