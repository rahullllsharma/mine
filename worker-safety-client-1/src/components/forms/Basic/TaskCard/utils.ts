import type { IconName } from "@urbint/silica";
import classnames from "classnames";
import { RiskLevel } from "@/api/generated/types";
import { RiskLevel as RiskBadgeRiskLevel } from "@/components/riskBadge/RiskLevel";

const convertRiskLevelToRiskBadgeRiskLevel = (
  riskLevel: RiskLevel
): RiskBadgeRiskLevel => {
  switch (riskLevel) {
    case RiskLevel.High:
      return RiskBadgeRiskLevel.HIGH;
    case RiskLevel.Medium:
      return RiskBadgeRiskLevel.MEDIUM;
    case RiskLevel.Low:
      return RiskBadgeRiskLevel.LOW;
    case RiskLevel.Unknown:
      return RiskBadgeRiskLevel.UNKNOWN;
    case RiskLevel.Recalculating:
      return RiskBadgeRiskLevel.RECALCULATING;
  }
};

const getBackgroundColorByRiskLevel = (risk: RiskLevel): string => {
  return classnames({
    "bg-risk-high": risk === RiskLevel.High,
    "bg-risk-medium": risk === RiskLevel.Medium,
    "bg-risk-low": risk === RiskLevel.Low,
    "bg-risk-unknown": risk === RiskLevel.Unknown,
    "bg-risk-recalculating": risk === RiskLevel.Recalculating,
  });
};

const getDefaultTextColorByRiskLevel = (risk: RiskLevel): string => {
  switch (risk) {
    case RiskLevel.High:
    case RiskLevel.Low:
      return "text-white";
    case RiskLevel.Medium:
      return "text-neutral-shade-100";
    case RiskLevel.Unknown:
    case RiskLevel.Recalculating:
      return "text-neutral-shade-75";
  }
};

const getRiskIcon = (risk?: RiskLevel): IconName | undefined => {
  switch (risk) {
    case RiskLevel.High:
      return "chevron_up";
    case RiskLevel.Medium:
      return "tilde";
    case RiskLevel.Low:
      return "chevron_down";
    default:
      return undefined;
  }
};

const getBorderColorByRiskLevel = (risk: RiskLevel): string => {
  return classnames({
    "border-risk-high": risk === RiskLevel.High,
    "border-risk-medium": risk === RiskLevel.Medium,
    "border-risk-low": risk === RiskLevel.Low,
    "border-risk-unknown": risk === RiskLevel.Unknown,
  });
};

export {
  convertRiskLevelToRiskBadgeRiskLevel,
  getBackgroundColorByRiskLevel,
  getDefaultTextColorByRiskLevel,
  getBorderColorByRiskLevel,
  getRiskIcon,
};
