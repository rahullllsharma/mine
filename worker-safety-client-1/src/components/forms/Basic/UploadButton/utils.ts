import type * as T from "fp-ts/lib/Task";

// This function assumes image resizing method as "contain" as this is our current desire
const drawImageToCanvas = (
  img: HTMLImageElement,
  width: number,
  height: number
): HTMLCanvasElement => {
  const canvas = document.createElement("canvas");
  const ctx = canvas.getContext("2d");

  canvas.width = width;
  canvas.height = height;

  ctx?.save();
  ctx?.drawImage(img, 0, 0, width, height);
  ctx?.restore();

  return canvas;
};

const generateFileObjectFromBase64 =
  (base64: string) =>
  (file: File): File => {
    const encodedBase64 = base64.split(",")[1];
    const encodedByteArr = atob(encodedBase64);

    let n = encodedByteArr.length;
    const u8arr = new Uint8Array(n);

    while (n--) {
      u8arr[n] = encodedByteArr.charCodeAt(n);
    }

    return new File([u8arr], file.name, { type: file.type });
  };

const readFileAsDataUrl = async (file: File): Promise<string> => {
  const result_base64 = await new Promise<string>(resolve => {
    const reader = new FileReader();
    reader.onload = () => resolve(reader.result as string);
    reader.readAsDataURL(file);
  });

  return result_base64;
};

const loadImage = async (base64: string): Promise<HTMLImageElement> => {
  const loadedImage = await new Promise<HTMLImageElement>(resolve => {
    const image = new Image();
    image.src = base64;
    image.onload = () => resolve(image);
  });

  return loadedImage;
};

const downscaleImage =
  (file: File) =>
  (maxFileSize: number): Promise<File> => {
    return new Promise<File>(async (resolve, _) => {
      if (file.size < maxFileSize) resolve(file);

      const readerResult = await readFileAsDataUrl(file);

      const image = await loadImage(readerResult);

      const { height: baseHeight, width: baseWidth } = image;

      let resizedHeight;
      let resizedWidth;

      if (baseHeight > baseWidth) {
        const targetHeight = 1080;
        resizedHeight = targetHeight;
        resizedWidth = baseWidth * (resizedHeight / baseHeight);
      } else {
        const targetWidth = 1080;
        resizedWidth = targetWidth;
        resizedHeight = baseHeight * (resizedWidth / baseWidth);
      }

      const resizedImageOnCanvas = drawImageToCanvas(
        image,
        resizedWidth,
        resizedHeight
      );

      const resizedImage = await generateFileObjectFromBase64(
        resizedImageOnCanvas.toDataURL()
      )(file);

      resolve(await resizedImage);
    });
  };

const downscaleImageAsTask =
  (file: File) =>
  (maxFileSize: number): T.Task<File> => {
    return () => downscaleImage(file)(maxFileSize);
  };

export { downscaleImage, downscaleImageAsTask };
