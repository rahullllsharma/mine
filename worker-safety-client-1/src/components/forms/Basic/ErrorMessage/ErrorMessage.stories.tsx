import type { ComponentMeta, ComponentStory } from "@storybook/react";
import type { Either } from "fp-ts/Either";
import type { Errors } from "io-ts";

import { NonEmptyString, withMessage } from "io-ts-types";
import { initFormField, updateFormField } from "@/utils/formField";

import { ErrorMessage } from "./ErrorMessage";

const decodeStringInput: (input: string) => Either<Errors, NonEmptyString> =
  withMessage(NonEmptyString, () => "Please select one option").decode;

let field = initFormField(decodeStringInput)("");
field = updateFormField(decodeStringInput)("");

export default {
  title: "JSB/ErrorMessage",
  component: ErrorMessage,
} as ComponentMeta<typeof ErrorMessage>;

export const Playground: ComponentStory<typeof ErrorMessage> = args => (
  <ErrorMessage {...args} />
);
Playground.args = {
  field: field,
};
