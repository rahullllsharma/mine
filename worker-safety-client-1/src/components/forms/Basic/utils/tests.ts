import type { Either } from "fp-ts/lib/Either";
import type { Errors } from "io-ts";
import * as t from "io-ts-types";
import { initFormField } from "@/utils/formField";

type DecodeFunction = (input: string) => Either<Errors, t.NonEmptyString>;

const createStringInputDecoder = (errorMessage: string) => {
  const decodeStringInput: DecodeFunction = t.withMessage(
    t.NonEmptyString,
    () => errorMessage
  ).decode;

  return decodeStringInput;
};

const createFormField = (raw: string, decodeFunction: DecodeFunction) => {
  return initFormField(decodeFunction)(raw);
};

export { createStringInputDecoder, createFormField };
