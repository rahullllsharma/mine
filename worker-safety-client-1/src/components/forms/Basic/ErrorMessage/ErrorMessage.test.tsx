import type { Either } from "fp-ts/lib/Either";
import type { Errors } from "io-ts";

import * as t from "io-ts-types";
import { render, screen } from "@testing-library/react";
import { initFormField, updateFormField } from "@/utils/formField";

import { ErrorMessage } from "./ErrorMessage";

describe(ErrorMessage.name, () => {
  const decodeStringInput: (input: string) => Either<Errors, t.NonEmptyString> =
    t.withMessage(t.NonEmptyString, () => "This field is required").decode;
  const createField = (raw = "") => {
    return initFormField(decodeStringInput)(raw);
  };

  it("renders correctly", () => {
    let field = createField();
    field = updateFormField(decodeStringInput)("");
    const { asFragment } = render(<ErrorMessage field={field} />);

    expect(screen.getByText("This field is required")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
