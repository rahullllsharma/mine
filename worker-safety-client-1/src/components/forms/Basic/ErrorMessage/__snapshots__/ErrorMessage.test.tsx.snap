// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorMessage renders correctly 1`] = `
<DocumentFragment>
  <span
    class="text-sm mt-1 text-system-error-40"
  >
    This field is required
  </span>
</DocumentFragment>
`;
